libtenzir/builtins/aggregation-functions/collect.cpp

```diff
@@ -47,13 +49,44 @@ class collect_function final : public aggregation_function {
   }
 
   auto finish() && -> caf::expected<data> override {
+    // XXX: Why exchange? Is this same instance reused?
     return data{std::exchange(result_, {})};
   }
 
   list result_ = {};
 };
 
-class plugin : public virtual aggregation_function_plugin {
+class collect_instance final : public aggregation_instance {
+public:
+  explicit collect_instance(ast::expression expr) : expr_{std::move(expr)} {
+  }
+
+  auto update(const table_slice& input, session ctx) -> void override {
+    auto arg = eval(expr_, input, ctx);
+    if (caf::holds_alternative<null_type>(arg.type)) {
+      return;
+    }
+    // XXX: What should happen if it receives different types?
+    // Currently, everything ends up coerced to strings.
```

Comment:
For some type combinations (e.g. `ip` and `duration`), we can't do much better, because we do not have union types (yet). For others, such as `int64` and `uint64`, we could attempt to unify the values, but that is probably out-of-scope for this PR.